src/util/interpolate.ts

```diff
@@ -1,6 +1,25 @@
+import {hclToRgb, labToRgb} from './color_spaces';
 import Color from './color';
 import Padding from './padding';
 
+export type InterpolationColorSpace = 'rgb' | 'hcl' | 'lab';
+
+/**
+ * Checks whether the specified color space is one of the supported interpolation color spaces.
+ *
+ * @param colorSpace Color space key to verify.
+ * @returns `true` if the specified color space is one of the supported
+ * interpolation color spaces, `false` otherwise
+ */
+export function isSupportedInterpolationColorSpace(colorSpace: string): colorSpace is InterpolationColorSpace {
+    return colorSpace === 'rgb' || colorSpace === 'hcl' || colorSpace === 'lab';
+}
+
+/**
+ * @param interpolationType Interpolation type
+ * @returns interpolation fn
+ * @deprecated use `interpolate[type]` instead
```

Comment:
I definitely like this factory better than the exported object and avoid exporting the object at the end?
Can we use it instead of `interpolate[type]`?
It's a lot easier to understand from the calling code what can be the options here.

Reply:
Ideally, this `interpolate` object should be exported as:
```ts
export default {
    number,
    color,
    array,
    padding,
};
```
but because this project uses `dts-bundle-generator` which does not handle multiple export default of unnamed object; that is why this rule exists:
https://github.com/maplibre/maplibre-style/blob/16fc6a7eb84ea5a3c963474da09b3477e666ceb4/.eslintrc#L83
all default exports must have a name. If it's because of this repetition `const obj = { ... }; export default obj;` that you don't like the exported object you might consider migrating from `dts-bundle-generator` to for example `rollup-plugin-dts` which, at least in my local attempts, did not have any problems with multiple modules/files with unnamed default object exports.


As for this "factory function"... No, the object must be exported. If you have any doubt then look for `interpolates.number` calls in maplibre-gl-js codebase and imagine how that code would look like if replaced with this `interpolateFactory`.

You may not be aware of it but from TS point of view this function looks like this:
```ts
declare function number(from: number, to: number, t: number): number;
declare function color(from: Color, to: Color, t: number, spaceKey?: InterpolationColorSpace): Color;
declare function array<T extends number[]>(from: T, to: T, t: number): T;
declare function padding(from: Padding, to: Padding, t: number): Padding;
declare const interpolateFactory: (interpolationType: "number" | "color" | "array" | "padding") =>
    typeof array | typeof number | typeof color | typeof padding;
```
I hope I do not have to explain to you why this is wrong and why it is a bad candidate for being part of a library public api.

And no, I don't agree that it's easier to understand what options you have by calling a function that takes the function name as a string, it's easier to understand what options you have by looking at the object and what fields that object has. From my point of view, it's like saying that `consoleFactory('log', 'hello')` is somehow better than `console.log('hello')`.

Reply:
The equivalent is calling `console[operationType].write("do something")` which for me is a shot in the dark when reading this code as there is no good way to know what `operationType` is and no good way to safe guard it, in the example I gave, yes, I prefer `consoleFactory('log').write('hello')` since I can easily understand what I get and what I can do when I read the code.

Reply:
There are plenty of good ways to find out what type `operationType` is, like TS types or type guards, and on top of that both approaches have the same level of protection against an invalid argument, both will simply return undefined :D

Consider the following example ([ts playground version](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABGEBbARgUwE4ApjZyoBcyaW2ANIlHKShjtVPeTgJSuPaIDeAUIiGJsmKCGxICRRAGoaiAFSJctRAFpE01OwDc-AL79+oSLASIAhtmyWAnvkIky3ANoBdZnRcUPzLhScPjgefILCouKSWk4AdKiWAA64uAAm1DDsiAC8AHzBeOk0cK4wnjTseobGEAgAzlCIMGBQOIlwADaWrTl8DBTU1rZ2Bvr8tWANTS1tnd2YAGKW0HDYdr24za3Y7V3mYAAqdomYpADk-ThniAA+iGdD9mdZeWHCiHUA7jBQEAAWKi2sz28EOx0wWQE73eEEsdUw90u2DOpEiEiQSP00OEsPh90edhRIjE6KsNnsWOERlGxgA9LTEOg4TAIIhMAAPSyoRIdTDjeqNIYARl6riF1AATO59BMpkMJaKJdQAGzS-mTRqNbKIAAMsQArGNZY1sCLtUCdnNWksVmtcA9yYT2LhhYMbEqKrpEPSaHU2TZVsR+D6AILYADmaEwLUQcGANHBiLY2A81xgfrAcEFdTqMHDYEs6F5xUQiWsXLEOFj8agiYuycQADICqnYsGGabqwmTgVbi33DUBSIFeaZpa9phYgSXTZxWTsB6oHpvQzaycM0M4J9MKlELUbJhoB07NRsABmLtrhFIjzqqbn3oW3bzVwO4Zndwz01uhfMZc+q8NwDbdd33UQjxPEQLzjbtr2TW922mAA3SwOhgXcC1QBEOS5Hk+T3IdmhQtDUgAOQrXoziI1D0LOGUh2wAAWR8x2fa1lloO1qJI8isKsP1LDAOwqmNER9RY7Y2MwUowGI9DeIROErCE6UVzZdkeRZH49zhRpVlgxBI2sXdRAARxAGBRFSO8TWVCTgRfbj5IopT624a47jfJ41X4KFhFMaBQWmSSrUWDjVgcJ8rVBI4TnOJFnlIK8YMxcIhAC-Zgoc9jbUi1jooQWLTnxR1EtgmCCUpdLwECiwoonG1OLykKQUK8F4uTDySvfSE0uhL4fn+QF8tasETl67FsVxBE3IoIk0WiVLJuhabuqeVESWiSq+veIxdr60TsAAdns8d5kaiL7QJZ4vzneU-y9H1M0+JpGnTWMAGtMDsaogA)):
```ts
const arr1 = [1, 2];
const arr2 = [2, 6];
const t = 0.5;
const r1 = interpolateFactory('array')(arr1, arr2, t); // ts error:
// Argument of type 'number[]' is not assignable to parameter of type 'Color & number[] & number & Padding'.
// r1 of type number | Color | number[] | Padding

const r2 = interpolate.array(arr1, arr2, t); // types narrowed correctly, r2 of type number[]
const r3 = interpolate['array'](arr1, arr2, t); // types narrowed correctly, r3 of type number[]
```
For this example to work properly the `interpolateFactory` function would need to be overloaded for each `interpolationType`:
```ts
export function interpolateFactory(interpolationType: 'number'): typeof number;
export function interpolateFactory(interpolationType: 'color'): typeof color;
export function interpolateFactory(interpolationType: 'array'): typeof array;
export function interpolateFactory(interpolationType: 'padding'): typeof padding;
export function interpolateFactory(interpolationType: 'number' | 'color' | 'array' | 'padding') {
    // switch case
}
```

Also, check out all the places where this factory function is used and see how it is used:
https://github.com/maplibre/maplibre-gl-js/blob/c62155b28818ac5777e4853cffe2892a0ae59487/src/style/properties.ts#L486-L491
https://github.com/maplibre/maplibre-gl-js/blob/c62155b28818ac5777e4853cffe2892a0ae59487/src/style/properties.ts#L545-L550

I don't mean to be intentionally rude, I'm just trying to share my own experience, and I assure you that it's much better to expose in [public API](https://github.com/maplibre/maplibre-style/blob/16fc6a7eb84ea5a3c963474da09b3477e666ceb4/src/style-spec.ts) interpolate object rather than `interpolateFactory` function. `switch case` is not always the answer to all problems! 😜